### DISCUSSION TITLE
How do I need to implement a One-to-many and many-to-one Relationship?

### DISCUSSION THREAD

**Discussion by NeoCortex97** [sqlalchemy/sqlalchemy#8653] [Usage Questions]
Hi, I started feeling stupid, when I tried to implement a simple, internal tool that deals with users and projects.
I wanted a User to have one project and a Project to have multiple users.

I don't know from the docs, if I implemented it right, but instead of explaining too much, I would like to show you a minimal version of my models, so you could correct me if I am wrong:

```py
class Project(Base):
    __tablename__ = "projects"
    users = Column(Integer, ForeignKey('users.id'))
```
```py
class User(Base):
    __tablename__ = 'users'
    project = relationship("Project")
```
I stripped all fields that are irrelevant to this problem to keep this post as short as possible...
I do know SQL, but I don't like working with raw queries, so I might be a little rusty on that topic.
My Intuition tells me that I should put the `ForeignKey` object in the `User`-model, but from the docs it seemed otherwise.

Maybe I overlooked a part of the docs, where the example was with `Parent` and `Child`. I thought that one Parent may only have one Child, but one child could have multiple parents.

Edit:
I want a bidirectional relationship between User and Project, where one User may only have one Project, but one project may have multiple users.

**Comment by zzzeek** [comment]
> 
> I don't know from the docs, if I implemented it right, but instead of explaining too much, I would like to show you a minimal version of my models, so you could correct me if I am wrong:
> 
> ```python
> class Project(Base):
>     __tablename__ = "projects"
>     users = Column(Integer, ForeignKey('users.id'))
> ```
> 
> ```python
> class User(Base):
>     __tablename__ = 'users'
>     project = relationship("Project")
> ```

OK so if one Project has multiple Users, think in SQL.    one row for Project, with id=5 let's say.  then many rows for User, id=1, id=2, id=3, ... which need to refer to Project.   They point to project, so they'd have something like a "project_id" column, and on "users.project_id" you'd configure ForeignKey to "projects.id", the primary key col of Project.

IOW, Project->User is one-to-many, "one" project for "many" users".   you'd follow the example at https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#one-to-many .

> 
> Maybe I overlooked a part of the docs, where the example was with `Parent` and `Child`. I thought that one Parent may only have one Child, but one child could have multiple parents.


the docs use Parent/Child in terms of the "one-to-many" or "many-to-one" where it would be "one (Parent) to many (Child)" or "many (Parent) to one (Child)", if that makes sense.   the symmetry thing makes it inherently tricky to not get confusing, myself included.

**Reply by NeoCortex97** [reply]
Tanks, that cleared some things for me.
But how can i set a project for the user?

my first guess would be the following:
```py
user.project_id == project.id
session.update(user)
session.commit()
```
Or is there a better way?

**Reply by zzzeek** [reply]
OK I would suggest walking through the [tutorial](https://docs.sqlalchemy.org/en/14/tutorial/orm_related_objects.html#working-with-related-objects) regarding this, which shows the basic idea is that you deal with the attribute that you put relationship() on, the "id" stuff is taken care of.

I just applied "Project" and "User" to my stock "relationship" example below.  You can run this program as is and study what it does.

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


class Project(Base):
    __tablename__ = "project"

    id = Column(Integer, primary_key=True)
    data = Column(String)
    users = relationship("User", back_populates="project")


class User(Base):
    __tablename__ = "user"
    id = Column(Integer, primary_key=True)
    project_id = Column(ForeignKey("project.id"))
    data = Column(String)
    project = relationship("Project", back_populates="users")


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

project1 = Project(data="some project")

u1 = User(data="user1")
u2 = User(data="user2")
u3 = User(data="user3")

# append users to projects...
project1.users.append(u1)
project1.users.append(u2)

# or assign project to user
u3.project = project1

# same effect, this will print out all three users
print(project1.users)

# just add the lead object to the Session
s.add(project1)

# commit - look at SQL output in console to see what happened
s.commit()

```
